perun/logic/runner.py

```diff
@@ -159,43 +195,25 @@ def run_all_phases_for(runner, runner_type, runner_params):
     :param module runner: module that is going to be runned
     :param str runner_type: string type of the runner (either collector or postprocessor)
     :param dict runner_params: dictionary of arguments for runner
+    :return RunnerReport: report about the run phase
     """
-    ok_status = CollectStatus.OK if runner_type == 'collector' else PostprocessStatus.OK
-    error_status = CollectStatus.ERROR if runner_type == 'collector' else PostprocessStatus.ERROR
     runner_verb = runner_type[:-2]
 
+    report = RunnerReport(runner, runner_type, runner_params)
+
     for phase in ['before', runner_verb, 'after']:
-        phase_function = getattr(runner, phase, None)
-        if phase_function:
-            try:
-                ret_val, ret_msg, updated_params = phase_function(**runner_params)
-            # We safely catch all of the exceptions
-            except Exception as exc:
-                ret_val, ret_msg, updated_params = error_status, str(exc), {}
-            runner_params.update(updated_params or {})
-            if not is_status_ok(ret_val, ok_status):
-                return error_status, "error while {}{} phase: {}".format(
-                    phase, ("_" + runner_verb)*(phase != runner_verb), ret_msg
-                ), None
-        elif phase == runner_verb:
-            return error_status, "missing {}() function for {}".format(
-                runner_verb, runner.__name__
-            ), None
-
-    # Return the processed profile
-    if 'profile' not in runner_params.keys():
-        return error_status, "missing generated profile for {} {}".format(
-            runner_type, runner.__name__
-        ), None
-    if not isinstance(runner_params['profile'], profile_factory.Profile):
-        log.warn("{} {} does not return Profile object".format(
-            runner_type, runner.__name__
-        ))
-        return ok_status, "", profile_factory.Profile(runner_params['profile'])
-    return ok_status, "", runner_params['profile']
+        run_phase_function(report, phase)
```

Comment:
I think that we're missing the signal handling that would trigger the `teardown` phase. While SIGINT would be caught with the KeyboardInterrupt exc., other termination signals would get past the exception handling in the `run_phase_function`, ignoring the cleanup code (such as SIGTERM or SIGQUIT, especially the SIGTERM is used from time to time). 

Also, this might be an almost negligible corner case, however, wouldn't the `teardown` phase be skipped if the SIGINT (KeyboardInterrupt) is raised outside of the `run_phase_function`, e.g. line 204 / 207 / 208?

Reply:
You are right, I forgot we were talking about the support of signals. Did you resolve similar issue in the past? Some snippet of code or page/article would be helpful for resolving this issue.